services_app/src/main/java/org/opendatakit/services/sync/service/logic/ProcessManifestContentAndFileChanges.java

```diff
@@ -960,6 +968,50 @@ public boolean syncRowLevelFileAttachments(String serverInstanceFileUri, String
     }
   }
 
+  /**
+   * Wrapper around downloadFile with the default maximum number
+   * of retries set to DEFAULT_DL_MAX_RETRY_COUNT
+   *
+   * @param destFile
+   * @param downloadUri
+   * @param expectedMd5Hash
+   * @return true if the download was successful, false if otherwise
+   * @throws IOException
+   */
+  private boolean downloadFile(File destFile, URI downloadUri, String expectedMd5Hash)
+      throws IOException {
+    return downloadFile(destFile, downloadUri, expectedMd5Hash, DEFAULT_DL_MAX_RETRY_COUNT);
+  }
+
+  /**
+   * Wrapper around Synchronizer.downloadFile that invokes that method first
+   * then checks the downloaded file's integrity.
+   *
+   * Negative maxRetry is considered as 0.
+   *
+   * @param destFile
+   * @param downloadUri
+   * @param expectedMd5Hash
+   * @param maxRetry
+   * @return true if the download was successful, false if otherwise
+   * @throws IOException
+   */
+  private boolean downloadFile(File destFile, URI downloadUri, String expectedMd5Hash, int maxRetry)
+      throws IOException {
+    if (maxRetry < 0) {
+      maxRetry = 0;
+    }
+
+    boolean hashMatch;
+
+    do {
+      sc.getSynchronizer().downloadFile(destFile, downloadUri);
+      hashMatch = ODKFileUtils.getMd5Hash(sc.getAppName(), destFile).equals(expectedMd5Hash);
+    } while (maxRetry-- > 0 && !hashMatch);
```

Comment:
If we want the retries to be limited to 3 times, then we should change the expression to 
--maxRetry > 0.

Reply:
The first download doesn't count as a retry.

Reply:
Fair enough.  I'll merge it.